stablehlo/dialect/TypeInference.cpp

```diff
@@ -3453,6 +3489,59 @@ LogicalResult verifyConvolutionOp(
                              "is incompatible with return type of operation ",
                              shapedResultType, "");
 
+  llvm::SmallVector<Type, 3> typeEntries{lhsType, rhsType, resultType};
+  if (noneQuantized(typeEntries)) return success();
+  // convolution_c28
+  if (!allQuantized(typeEntries)) {
+    return emitOptionalError(location,
+                             "not all of operands and result are quantized");
+  }
+
+  auto lhsQType =
+      getElementTypeOrSelf(lhsType).dyn_cast<quant::QuantizedType>();
+  auto rhsQType =
+      getElementTypeOrSelf(rhsType).dyn_cast<quant::QuantizedType>();
+  auto resultQType =
+      getElementTypeOrSelf(resultType).dyn_cast<quant::QuantizedType>();
+  // convolution_c29
+  if (lhsQType.getStorageType() != rhsQType.getStorageType())
+    return emitOptionalError(location, "mismatched operand storage types ",
+                             lhsQType.getStorageType(), " and ",
+                             rhsQType.getStorageType(), "");
+  // convolution_c30
+  auto expressedType = lhsQType.getExpressedType();
+  if (expressedType != rhsQType.getExpressedType() ||
+      expressedType != resultQType.getExpressedType())
+    return emitOptionalError(location,
+                             "mismatched operands and result expressed types");
+
+  llvm::SmallVector<Type, 2> typeEntriesPerAxis{rhsType, resultType};
+  if (nonePerAxisQuantized(typeEntriesPerAxis)) return success();
+  // convolution_c31
+  if (!allPerAxisQuantized(typeEntriesPerAxis)) {
```

Comment:
Why is this an error? we can have `lhs` per-tensor and `rhs` per-axis, right?

Reply:
(C31) says if rhs is per-axis then result must be per-axis.

Reply:
we are not checking lhs here. typeEntriesPerAxis has {rhs, result} only